flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerDriverTest.java

```diff
@@ -1,4 +1,598 @@
+					ApplicationId.newInstance(System.currentTimeMillis(), 1),
+					1),
+				0),
+				NodeId.newInstance("container", 1234),
+				testingResource,
+				Priority.newInstance(1));
+		final TaskExecutorProcessSpec testingTaskExecutorProcessSpec =
+			new TaskExecutorProcessSpec(
+				new CPUResource(1),
+				MemorySize.ZERO,
+				MemorySize.ZERO,
+				MemorySize.ofMebiBytes(256),
+				MemorySize.ofMebiBytes(256),
+				MemorySize.ofMebiBytes(256),
+				MemorySize.ofMebiBytes(256),
+				MemorySize.ZERO,
+				MemorySize.ZERO);
+		private final CompletableFuture<String> stopAndCleanupClusterFuture =  new CompletableFuture<>();
+		private final CompletableFuture<Resource> createTaskManagerContainerFuture = new CompletableFuture<>();
+		private final List<CompletableFuture<Void>> addContainerRequestFutures = new ArrayList<>();
+		private final AtomicInteger addContainerRequestFuturesNumCompleted = new AtomicInteger(0);
+		private final CompletableFuture<Void> removeContainerRequestFuture = new CompletableFuture<>();
+		private final CompletableFuture<Void> releaseAssignedContainerFuture = new CompletableFuture<>();
+		private final CompletableFuture<Void> startContainerAsyncFuture = new CompletableFuture<>();
+		private final CompletableFuture<Void> stopContainerAsyncFuture = new CompletableFuture<>();
+
+		final TestingYarnNMClientAsync.Builder testingYarnNMClientAsyncBuilder = TestingYarnNMClientAsync.builder();
+		final TestingYarnAMRMClientAsync.Builder testingYarnAMRMClientAsyncBuilder = TestingYarnAMRMClientAsync.builder();
+
+		final Map<String, String> env = new HashMap<>();
+
+		int containerIdx = 0;
+
+		@Override
+		protected void prepareRunTest() {
+			File root = folder.getRoot();
+			File home = new File(root, "home");
+			env.put(ENV_APP_ID, "foo");
+			env.put(ENV_CLIENT_HOME_DIR, home.getAbsolutePath());
+			env.put(ENV_CLIENT_SHIP_FILES, "");
+			env.put(ENV_FLINK_CLASSPATH, "");
+			env.put(ENV_HADOOP_USER_NAME, "foo");
+			env.put(FLINK_DIST_JAR, new YarnLocalResourceDescriptor(
+				"flink.jar",
+				new Path("/tmp/flink.jar"),
+				0,
+				System.currentTimeMillis(),
+				LocalResourceVisibility.APPLICATION,
+				LocalResourceType.FILE).toString());
+			env.put(ApplicationConstants.Environment.PWD.key(), home.getAbsolutePath());
+		}
+
+		@Override
+		protected void preparePreviousAttemptWorkers() {
+			testingYarnAMRMClientAsyncBuilder.setRegisterApplicationMasterFunction((ignored1, ignored2, ignored3) -> RegisterApplicationMasterResponse.newInstance(
+				Resource.newInstance(0, 0),
+				Resource.newInstance(Integer.MAX_VALUE, Integer.MAX_VALUE),
+				Collections.emptyMap(),
+				null,
+				Collections.singletonList(testingContainer),
+				null,
+				Collections.emptyList()));
+		}
+
+		@Override
+		protected void prepareDeregisterApplication() {
+			testingYarnAMRMClientAsyncBuilder.setUnregisterApplicationMasterConsumer(
+				(ignore1, ignore2, ignore3) -> stopAndCleanupClusterFuture.complete("finish"));
```

Comment:
What is the purpose for `"finish"`? I think `CompletableFuture<Void>` should be good enough.